vowpalwabbit/csv_parser/src/parse_example_csv.cc

```diff
@@ -0,0 +1,280 @@
+  size_t num_bytes_consumed = read_line(all, ae, buf);
+  // Read the data if header exists.
+  if (!all->no_header && first_read) { num_bytes_consumed += read_line(all, ae, buf); }
+  return static_cast<int>(num_bytes_consumed);
+}
+
+size_t parser::read_line(VW::workspace* all, VW::example* ae, io_buf& buf)
+{
+  char* line = nullptr;
+  size_t num_chars_initial = buf.readto(line, '\n');
+  // This branch will get hit when we haven't reached EOF of the input device.
+  if (num_chars_initial > 0)
+  {
+    size_t num_chars = num_chars_initial;
+    if (line[0] == '\xef' && num_chars >= 3 && line[1] == '\xbb' && line[2] == '\xbf')
+    {
+      line += 3;
+      num_chars -= 3;
+    }
+    if (num_chars > 0 && line[num_chars - 1] == '\n') { num_chars--; }
+    if (num_chars > 0 && line[num_chars - 1] == '\r') { num_chars--; }
+
+    VW::string_view csv_line(line, num_chars);
+
+    if (csv_line.empty()) { ae->is_newline = true; }
+    else
+    {
+      std::vector<VW::string_view> elements = split(csv_line, all->separator);
+      // If no header present will number features as 1..k based on column number
+      if (all->no_header && _header.empty())
+      {
+        for (size_t i = 1; i < elements.size() + 1; i++) { _header.emplace_back(std::to_string(i)); }
+      }
+
+      if (!_header.empty() && elements.size() != _header.size())
+      {
+        THROW("CSV line has " << elements.size() << " elements, but the header has " << _header.size() << " elements.");
+      }
+      else if (!all->no_header && _header.empty())
+      {
+        for (size_t i = 0; i < elements.size(); i++)
+        {
+          check_if_float(elements[i]);
+          _header.emplace_back(elements[i]);
+        }
+      }
+      else
+      {
+        parse_example(all, ae, elements);
+      }
+    }
+  }
+  return num_chars_initial;
+}
+
+void parser::parse_example(VW::workspace* all, VW::example* ae, std::vector<VW::string_view> csv_line)
+{
+  parse_label(all, ae, csv_line);
+  // TODO: parse_tag(all, ae, csv_line);
+  parse_namespaces(all, ae, csv_line);
+}
+
+void parser::parse_label(VW::workspace* all, VW::example* ae, std::vector<VW::string_view> csv_line)
+{
+  all->example_parser->lbl_parser.default_label(ae->l);
+
+  // negative numbers support the "from end of line" convention
+  // (e.g: -1 is last column, -2 is next-to-last, etc.)
+  if (all->label < 0) { label_index = csv_line.size() + all->label; }
+  else
+  {
+    label_index = all->label;
+  }
+
+  if (label_index >= csv_line.size() || label_index < 0) { THROW("Label index out of range!"); }
+
+  VW::string_view label_content(csv_line[label_index]);
+
+  // Multiclass labels will be auto-converted to 1..k if they are
+  // non-numeric e.g. Species: {setosa, versicolor, virginica} -> {1, 2, 3}
+  std::string label_string;
+  if (!check_if_float(label_content))
+  {
+    auto it = multiclass_label_counter.find(label_content);
```

Comment:
Rather than rebuild this functionality, I think we should rely on the existing `--named_labels` functionality for this if possible

Reply:
Oh cool, didn't noticed that functionality before.

For the multiclass_label_counter, I guess it can also get removed now, too.

> This seems a bit dangerous. What is the lifetime of these string views? Is it possible for them to become invalid and still accessible in this map?

Reply:
Hi! I can't get the --named_labels working with `vw -d Iris.txt --named_labels Setosa,Versicolor,Virginica`

Am I get something wrong or the upstream has been broken?

[Iris.txt](https://github.com/HollowMan6/vowpal_wabbit/files/8675728/Iris.txt)

![image](https://user-images.githubusercontent.com/43995067/168011727-563c9533-8e47-433e-bc78-c78500e8dbb8.png)

If this is working, then just remove my realization would be OK and resolve this.

Reply:
Also maybe Multiclass labels auto-convert to 1..k would be better so that users do not need to manually specify that?

Reply:
The issue with auto-converting is that it depends on the order of the data. Presumably you assign an id each time you see a new class. That means that if the data is ordered a different way, a completely different model would be learned. Perhaps this is okay? Seems a bit scary to me

Reply:
> The issue with auto-converting is that it depends on the order of the data. Presumably you assign an id each time you see a new class. That means that if the data is ordered a different way, a completely different model would be learned. Perhaps this is okay? Seems a bit scary to me

Oh! Yes, that's really a issue to consider. But I can't get the `--named_labels` working so far, could you tell me if I did anything wrong here? If it's actually working, them later I will remove my auto-converting realization, but currently I would still remain that in this repo for testing purpose.

Reply:
Okay, cool. I didn't know that you have to also specify `--oaa 3` here to get it working!